### PR TITLE
fix(o3r): replaced path.resolve with path.join for better path resolution

### LINKED ISSUE #1778: [Bug]: Unable to create service using the Generator
### Package name

core

### Package version

10.2.4

### Reproduction steps

Create a simple NX repo
Add o3r/core to the repo
Add an Nx Angular Library for service
Try creating a service in the newly created library using command

"nx generate @o3r/core:service"

### Current result

After accepting the inputs, the process hanges

image

### Expected result

Service should be created

### Additional comments

Did a bit of digging around in the schematics.
 
It never exits out of the while loop in [Index.ts Line 49](https://github.com/AmadeusITGroup/otter/blob/main/packages/%40o3r/core/schematics/service/index.ts#L49)

image

"tree.exists" isnt working correctly.  The third path in the below screenshot exists, but the "tree exists check" is saying false. 

image

also path.resolve(destination, rec) !== '/' doesnt work in windows as it ends up returning directory like "C:\"
image

Issue appears to be command path.resolve

### PR DESCRIPTION
**Pull request by gavadhany**
## Proposed change
Change path resolver from path.resolve to path.join  as the tree.exists is unable to resolve the absolute path generated by path.resolve

## Related issues

- :bug: Fixes #1778

### CODE REVIEW

File: packages/@o3r/core/schematics/service/index.ts [hunk 1]
```diff
@@ -46,10 +46,10 @@ function ngGenerateServiceFn(options: NgGenerateServiceSchematicsSchema): Rule {
     let packageName = destination;
     if (workspaceProject?.projectType !== 'application') {
       let rec = '..';
-      while (!tree.exists(path.resolve(destination, rec, 'package.json')) && path.resolve(destination, rec) !== '/') {
+      while (!tree.exists(path.join(destination, rec, 'package.json')) && path.resolve(destination, rec) !== '/') {
```
Comment (matthieu-crouzet) [inline packages/@o3r/core/schematics/service/index.ts:49]:
```suggestion
      while (!tree.exists(path.join(destination, rec, 'package.json')) && path.join(destination, rec) !== '/') {
```

Comment (gavadhany) [reply packages/@o3r/core/schematics/service/index.ts]: @matthieu-crouzet can't use  "path.join" while comparing with root path as we need absolute path. Path.join returns relative path. Hence using path.resolve

Comment (matthieu-crouzet) [reply packages/@o3r/core/schematics/service/index.ts]: But you are comparing with the tree path that is a relative path to the project 
on Windows you will never have a path matching `'/'`

Comment (gavadhany) [reply packages/@o3r/core/schematics/service/index.ts]: Thats correct.. so made changes to accept both "Drive letter" and "/" in todays [commit](https://github.com/AmadeusITGroup/otter/compare/f97502a894f140f7f559a34a173703c218b6bac5..771e53a3031a92b7d84421ec93e74a7123d613c8)

Comment (gavadhany) [reply packages/@o3r/core/schematics/service/index.ts]: Cool.. will take a look prevent it from resolving paths outside projects

**Review by kpanot — APPROVED** [review]

**Comment by gavadhany** [comment]
@matthieu-crouzet and @kpanot : I have refactored the code a bit. The current version of the code wouldnt work if package.json was in the path entered by user e.g. './'. 
Also added a check to prevent path resolution outside the project workspace and an error package.json doesnt exist within the project workspace.

File: packages/@o3r/core/schematics/service/index.ts [hunk 2]
```diff
@@ -45,11 +44,16 @@ function ngGenerateServiceFn(options: NgGenerateServiceSchematicsSchema): Rule {
     const workspaceConfig = getWorkspaceConfig(tree);
     let packageName = destination;
     if (workspaceProject?.projectType !== 'application') {
-      let rec = '..';
-      while (!tree.exists(path.resolve(destination, rec, 'package.json')) && path.resolve(destination, rec) !== '/') {
-        rec = path.join('..', rec);
+      let currentDirectory = path.join(destination);
+      let parent = '..';
```
Comment (kpanot) [inline packages/@o3r/core/schematics/service/index.ts:48]: the value never change, it should be `const`.
(I am surprised it has not been caught by the linter :S)

File: packages/@o3r/core/schematics/service/index.ts [hunk 3]
```diff
@@ -45,11 +44,16 @@ function ngGenerateServiceFn(options: NgGenerateServiceSchematicsSchema): Rule {
...
+      // Find the package.json file in the project directory
+      while (currentDirectory !== parent && !tree.exists(path.join(currentDirectory, 'package.json'))) {
+        currentDirectory = path.join(currentDirectory, parent);
+      }
+      if (currentDirectory === parent) {
+        throw new SchematicsException('Could not find package.json in the project directory.');
```
Comment (matthieu-crouzet) [inline packages/@o3r/core/schematics/service/index.ts:54]:
```suggestion
        throw new O3rCliError('Could not find package.json in the project directory.');
```
from `@o3r/schematics`

File: packages/@o3r/core/schematics/service/index.ts [hunk 4]
```diff
@@ -45,11 +44,16 @@ function ngGenerateServiceFn(options: NgGenerateServiceSchematicsSchema): Rule {
     const workspaceConfig = getWorkspaceConfig(tree);
     let packageName = destination;
     if (workspaceProject?.projectType !== 'application') {
-      let rec = '..';
-      while (!tree.exists(path.resolve(destination, rec, 'package.json')) && path.resolve(destination, rec) !== '/') {
-        rec = path.join('..', rec);
+      let currentDirectory = path.join(destination);
+      let parent = '..';
+      // Find the package.json file in the project directory
+      while (currentDirectory !== parent && !tree.exists(path.join(currentDirectory, 'package.json'))) {
```
Comment (kpanot) [inline packages/@o3r/core/schematics/service/index.ts:50]: did you test that `currentDirectory !== parent` is correctly executing in the different OS?
It seems to be quite weak condition, maybe a `.include(parent)` would be a bit stronger?

File: packages/@o3r/core/schematics/service/index.ts [hunk 5]
```diff
@@ -45,11 +44,16 @@ function ngGenerateServiceFn(options: NgGenerateServiceSchematicsSchema): Rule {
     const workspaceConfig = getWorkspaceConfig(tree);
     let packageName = destination;
     if (workspaceProject?.projectType !== 'application') {
-      let rec = '..';
-      while (!tree.exists(path.resolve(destination, rec, 'package.json')) && path.resolve(destination, rec) !== '/') {
-        rec = path.join('..', rec);
+      let currentDirectory = path.join(destination);
```
[the remark this answers is not in the document]
Comment (gavadhany) [reply packages/@o3r/core/schematics/service/index.ts]: Yea.. wanted to normalize. Replaced it with normalize

File: packages/@o3r/core/schematics/service/index.ts (hunk 2 above)
Comment (gavadhany) [reply packages/@o3r/core/schematics/service/index.ts]: Fixed

File: packages/@o3r/core/schematics/service/index.ts (hunk 3 above)
Comment (gavadhany) [reply packages/@o3r/core/schematics/service/index.ts]: done

File: packages/@o3r/core/schematics/service/index.ts (hunk 4 above)
Comment (gavadhany) [reply packages/@o3r/core/schematics/service/index.ts]: @kpanot : Changed it from === to includes

**Review by cpaulve-1A — APPROVED** [review]

**Review by kpanot — APPROVED** [review]

**Pull request merged**